django/db/models/sql/compiler.py

```diff
@@ -1500,19 +1509,43 @@ def _get_field_choices():
 
     def get_converters(self, expressions):
         converters = {}
-        for i, expression in enumerate(expressions):
-            if expression:
+        slice_converters = {}
+        i = 0
+
+        for expression in expressions:
+            if isinstance(expression, Cols):
+                # Include the regular converters for the composite fields.
+                cols = expression.get_source_expressions()
+                cols_converters = self.get_converters(cols)
+                for j, (convs, col) in cols_converters.items():
+                    converters[i + j] = (convs, col)
+                # Include a slice converter to convert the composite fields to a tuple.
+                cols_len = len(expression)
+                slice_pos = (i, i + cols_len)
+                slice_converters[slice_pos] = ((Cols.db_converter,), expression)
+                i += cols_len
+            elif expression:
                 backend_converters = self.connection.ops.get_db_converters(expression)
                 field_converters = expression.get_db_converters(self.connection)
                 if backend_converters or field_converters:
                     converters[i] = (backend_converters + field_converters, expression)
+                i += 1
+            else:
+                i += 1
+
+        if slice_converters:
+            # Apply slice converters in reverse to avoid IndexErrors.
+            converters.update(reversed(slice_converters.items()))
```

Comment:
Can you explain how we could get `IndexError` here in more detail? Or is this no longer relevant?

Reply:
For example, we have a list `[1, 2, 3, 4, 5, 6]` and we want to convert some slices to tuples e.g. `[(1, 2), (3, 4), (5, 6)]` (the slices are `(0, 2), (2, 4), (4, 6)`).

We need to convert in reverse order.
1. `[1, 2, 3, 4, (5, 6)]` (slice `(4, 6)`)
2. `[1, 2, (3, 4), (5, 6)]` (slice `(2, 4)`)
3. `[(1, 2), (3, 4), (5, 6)]` (slice `(0, 2)`)

Maybe using the converters for this isn't the best idea, we could add a new function to improve readability.

Reply:
Would making everything mutually exclusive slice as opposed to a mix of indices and slices (even they are just singletons) simplify the logic anyhow?

Reply:
```
Would making everything mutually exclusive slice as opposed to a mix of indices and slices (even they are just singletons) simplify the logic anyhow?
```

I think so. It would also add an extra iteration which is why I was reluctant to do it.

```
can we expand the converters to be per components of the `ColPairs`
```

Yes, it's already expanded like that, the question is "what's the best way to wrap the fields in tuples" (which is was `ColPairs.db_converter` is for).

Reply:
> Yes, it's already expanded like that, the question is "what's the best way to wrap the fields in tuples?" (which is what ColPairs.db_converter is for).

Is that even required though? We don't we assign attributes to the underlying fields/columns that the composite field is composed of and let `CompositeAttribute.__get__` return tuples of them if necessary? I don't understand why assigning tuples is necessary. Is that to support annotations of composite field references that cannot be _spread_ this way?

Reply:
Yes, it's required when retrieving composite fields from the database, e.g. `.values("pk")` or `.value_list("pk")`.
This will issue a `SELECT tenant_id, id` query which we need to map to `pk`.

Reply:
>  This will issue a SELECT tenant_id, id query which we need to map to pk.

I could be missing something why are we mapping `slice(0, 1)` to `pk` when we could be mapping `0` to `tenant_id` and `1` to `id`. Do you see what I mean? Composite fields are entirely virtual so mapping to their parts instead should have the same effect.

Reply:
Yes you're correct that works for the `ModelIterable`, but not for the `ValuesIterable`, `ValuesListIterable` which don't return models.

Reply:
I have separated this into the `composite_fields_to_tuples` function.